### ISSUE TITLE
Deprecated field should not be warned against when it must be used during struct initialization

### ISSUE THREAD

**Issue by Ploppz** [rust-lang/rust#102777]
Given the following code: 

```rust
pub struct Foo {
    #[deprecated(since = "0.52.16", note = "use some other field instead")]
    pub is_project: bool
}

fn main() {
    let _ = Foo {
        is_project: false
    };
}
```

The current output is:

```
warning: use of deprecated field `Foo::is_project`: use some other field instead
 --> src/main.rs:8:9
  |
8 |         is_project: false
  |         ^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(deprecated)]` on by default
```

Ideally there should be no warning. Because it's nice to be allowed to use the simple way of constructing a struct, in which case there's no point in avoiding initializing deprecated structs to some default values, or am I the only one who thinks so? I just don't see how this warning can be useful in this use case.

**Comment by yuyang-ok** [comment]
Yes,I think should be no warning too.

**Comment by T4r4sB** [comment]
I think, in pattern-matching this warning should be disabled too:
```

match event {
   ...
    WindowEvent::MouseInput { device_id, state, button, modifiers: _ } => {
```
result:
```

use of deprecated field `winit::event::WindowEvent::MouseInput::modifiers`: Deprecated in favor of WindowEvent::ModifiersChanged
`#[warn(deprecated)]` 

```

**Comment by dhruvja** [comment]
Just face this issue and was wondering if this is implemented?

**Comment by amitu** [comment]
If you are facing this issue you can do:

```rs
pub struct Foo {
    #[deprecated(since = "0.52.16", note = "use some other field instead")]
    pub is_project: bool
}

#[allow(deprecated)]
pub fn foo() {
    let _f = Foo { is_project: true };
}
```

Allow using a deprecated `item` internally in the crate without warning? Maybe an `#[externally_deprecated()]` attribute, so you do not have to do `#[allow(deprecated)]` all over the place while you are in the process of removing deprecated stuff? If I was maintaining a crate I would prefer the explicit `#[allow(deprecated)]` as that acts as TODO for me, I have to get rid of my own usage of deprecated stuff before users of my crate can do so.

**Comment by Enselic** [comment]
Triage: Hi! Isn't this a duplicate of https://github.com/rust-lang/rust/issues/47219? If not, why not?
